File: 프로그래머스/2/42890. 후보키/후보키.js
function solution(relation) {
    const rowLen = relation.length;      // 튜플(행) 개수
    const colLen = relation[0].length;   // 속성(열) 개수
    const candidates = [];               // 후보키 목록

    // 1) 가능한 모든 속성 조합(부분집합)을 비트마스크로 생성
    for (let bit = 1; bit < (1 << colLen); bit++) {
        const set = [];

        // 해당 비트 조합이 어떤 컬럼을 포함하는지 체크
        for (let col = 0; col < colLen; col++) {
            if (bit & (1 << col)) set.push(col);
        }

        // 2) 유일성 검사
        const tuples = new Set();
        for (let row = 0; row < rowLen; row++) {
            let key = set.map(col => relation[row][col]).join(',');
            tuples.add(key);
        }

        if (tuples.size !== rowLen) continue; // 유일성 실패

        // 3) 최소성 검사
        let minimal = true;
        for (const c of candidates) {
            // 기존 후보키와 비교: 기존 후보키가 현재 조합의 부분집합이면 최소성 위반
            if ((c & bit) === c) {
                minimal = false;
                break;
            }
        }

        if (minimal) candidates.push(bit);
    }

    return candidates.length;
}

File: 프로그래머스/2/42890. 후보키/후보키.py
from itertools import combinations

def solution(relation):
    row_len = len(relation)
    col_len = len(relation[0])
    candidate_keys = []

    # 1) 1개 컬럼부터 전체 컬럼까지 모든 조합 탐색
    for r in range(1, col_len + 1):
        for cols in combinations(range(col_len), r):

            # 2) 유일성 검사
            seen = set()
            for row in relation:
                key = tuple(row[c] for c in cols)
                seen.add(key)

            if len(seen) != row_len:
                continue  # 유일성 FAIL

            # 3) 최소성 검사
            minimal = True
            for ck in candidate_keys:
                # ck가 cols의 부분집합이면 최소성 위배
                if set(ck).issubset(set(cols)):
                    minimal = False
                    break

            if minimal:
                candidate_keys.append(cols)

    return len(candidate_keys)
